File: catkin_ws/src/spring2016/f4-devel/kinematics/src/inverse_kinematics_node.cpp
//
// Created by teddy on 3/17/16.
//
#include "ros/ros.h"
#include <string>
#include <duckietown_msgs/Twist2DStamped.h>
#include <duckietown_msgs/WheelsCmdStamped.h>

using namespace std;

class inverse_kinematics_node
{
public:
  inverse_kinematics_node(); // constructor

// class variables_
private:
  string node_name_;
  ros::NodeHandle nh_;
  string veh_name_;

  ros::Subscriber sub_twist_;
  ros::Publisher pub_wheelsCmd_;

  double K_l_; // duty to left wheel rotation rate constant
  double K_r_; // duty to right wheel rotation rate constant
  double radius_l_; // radius of the left wheel
  double radius_r_; // radius of the right wheel
  double baseline_lr_; //distance between the center of the two wheels

  void carCmdCallback(duckietown_msgs::Twist2DStampedConstPtr const& msg);
};

int main(int argc, char **argv){
  ros::init(argc, argv, "inverse_kinematics_node");
  inverse_kinematics_node node;
  ros::spin();
  return 0;
}

inverse_kinematics_node::inverse_kinematics_node(): nh_("~"), node_name_("inverse_kinematics_node")
{
  //Setup the publishers and subscirbers
  pub_wheelsCmd_ = nh_.advertise<duckietown_msgs::WheelsCmdStamped>("wheels_cmd", 10);
  sub_twist_ = nh_.subscribe("car_cmd", 10, &inverse_kinematics_node::carCmdCallback, this);

  //Get parameters
  nh_.param("K_l", K_l_, 25.0);
  nh_.param("K_r", K_r_, 25.0);
  nh_.param("radius_l", radius_l_, 0.02);
  nh_.param("radius_r", radius_r_, 0.02);
  nh_.param("baseline_lr", baseline_lr_, 0.1);

  ROS_INFO_STREAM("[" << node_name_ << "] has started.");
}

void inverse_kinematics_node::carCmdCallback(duckietown_msgs::Twist2DStampedConstPtr const& msg)
{
  // conversion from linear and angular velocities to motor rotation rate
  double omega_r = (msg->v + 0.5 * msg->omega * baseline_lr_)/radius_r_;
  double omega_l = (msg->v - 0.5 * msg->omega * baseline_lr_)/radius_l_;

  // conversion from motor rotation rate to duty cycle
  double u_r = omega_r / K_r_;
  double u_l = omega_l / K_l_;

  // stuff the wheel commands in a message and publish
  duckietown_msgs::WheelsCmdStamped cmd_msg;
  cmd_msg.header = msg->header;

  cmd_msg.header.stamp = ros::Time::now();  //Keep the time the command was given to the wheels_driver
  cmd_msg.vel_left = float(u_l);
  cmd_msg.vel_right = float(u_r);
  pub_wheelsCmd_.publish(cmd_msg);
}

File: catkin_ws/src/dagu_car/script/inverse_kinematics_node.py
#!/usr/bin/env python
import rospy
from duckietown_msgs.msg import WheelsCmdStamped, Twist2DStamped
from numpy import *


# Inverse Kinematics Node
# Author: Robert Katzschmann

class InverseKinematicsNode(object):
    def __init__(self):
        self.node_name = 'inverse_kinematics_node'

        # Setup the publisher and subscriber
        self.sub_car_cmd = rospy.Subscriber("~car_cmd", Twist2DStamped, self.car_cmd_callback)
        self.pub_wheels_cmd = rospy.Publisher("~wheels_cmd", WheelsCmdStamped, queue_size=1)

        # Read in parameters
        self.k_l = self.setup_parameter('k_l', 25.0)
        self.k_r = self.setup_parameter('k_r', 25.0)
        self.radius_l = self.setup_parameter('radius_l', 0.02)
        self.radius_r = self.setup_parameter('radius_r', 0.02)
        self.baseline = self.setup_parameter('baseline', 0.1)

        rospy.loginfo("[%s] has started", self.node_name)

    def car_cmd_callback(self, msg_car_cmd):
        # conversion from linear and angular velocities to motor rotation rate
        omega_r = (msg_car_cmd.v + 0.5 * msg_car_cmd.omega * self.baseline) / self.radius_r
        omega_l = (msg_car_cmd.v - 0.5 * msg_car_cmd.omega * self.baseline) / self.radius_l

        # conversion from motor rotation rate to duty cycle
        u_r = omega_r / self.k_r
        u_l = omega_l / self.k_l

        # Put the wheel commands in a message and publish
        msg_wheels_cmd = WheelsCmdStamped()
        msg_wheels_cmd.header = msg_car_cmd.header
        msg_wheels_cmd.vel_right = u_r
        msg_wheels_cmd.vel_left = u_l
        self.pub_wheels_cmd.publish(msg_wheels_cmd)

    def setup_parameter(self, param_name, default_value):
        value = rospy.get_param(param_name, default_value)
        # Write to parameter server for transparency
        rospy.set_param(param_name, value)
        rospy.loginfo("[%s] %s = %s " % (self.node_name, param_name, value))
        return value

if __name__ == '__main__':
    rospy.init_node('inverse_kinematics_node', anonymous=False)
    inverse_kinematics_node = InverseKinematicsNode()
    rospy.spin()
